servicehandler/servicediscovery.go

```diff
@@ -0,0 +1,209 @@
+			protocol: string(port.Protocol),
+		})
+	}
+}
+
+func (sra *ServiceAuthentication) Discover(ctx context.Context, scansWg *sync.WaitGroup, antsPool *ants.Pool, client dynamic.NamespaceableResourceInterface) {
+
+	// get all ports , each port equal different address
+	for _, pr := range sra.spec.ports {
+		if slices.Contains(protocolFilter, string(pr.protocol)) {
+			continue
+		}
+
+		//use DNS name to scan - this is the most reliable way to scan
+		srvDnsName := sra.metadata.name + "." + sra.metadata.namespace
+
+		scansWg.Add(1)
+		antsPool.Submit(func() {
+			pr.Scan(ctx, srvDnsName)
+			sra.spec.ports = append(sra.spec.ports, pr)
+			scansWg.Done()
+		})
+	}
+
+	serviceObj, structuredErr := sra.Unstructured()
+	if structuredErr != nil {
+		logger.L().Ctx(ctx).Error(structuredErr.Error())
+		return
+	}
+
+	_, deleteErr := client.Namespace(sra.metadata.namespace).Apply(context.TODO(), sra.metadata.name, serviceObj, metav1.ApplyOptions{FieldManager: FieldManager})
+	if deleteErr != nil {
+		logger.L().Ctx(ctx).Error(deleteErr.Error())
+	}
+}
+
+func (port *Port) Scan(ctx context.Context, ip string) {
+	result, err := cmd.ScanTargets(ctx, ip, port.port)
+	port.applicationLayer = result.ApplicationLayer
+	port.presentationLayer = result.PresentationLayer
+	port.sessionLayer = result.SessionLayer
+	port.authenticated = result.IsAuthenticated
+
+	if result.ApplicationLayer == "" {
+		// if we can't get the application layer, then we change to Unknown
+		port.applicationLayer = "Unknown"
+		port.authenticated = true
+	}
+
+	if err != nil {
+		//if we have an error, we log it and set all layers to Unknown
+		logger.L().Ctx(ctx).Error(err.Error())
+		result.ApplicationLayer = "Unknown"
+		result.PresentationLayer = "Unknown"
+		result.SessionLayer = "Unknown"
+		result.IsAuthenticated = true
+	}
+}
+
+func (csl currentServiceList) deleteServices(ctx context.Context, client dynamic.NamespaceableResourceInterface) error {
+	// get all services from the current cycle and compare them with the current CRDs
+
+	authServices, err := client.List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, service := range authServices.Items {
+		if !csl.contains(service.GetName(), service.GetNamespace()) {
+			err := client.Namespace(service.GetNamespace()).Delete(context.TODO(), service.GetName(), metav1.DeleteOptions{})
+			if err != nil {
+				logger.L().Error(err.Error())
+				continue
+			}
+			logger.L().Ctx(ctx).Info("Authentication Service " + service.GetName() + " in namespace " + service.GetNamespace() + " deleted")
+		}
+	}
+	return err
+}
+
+func DiscoveryServiceHandler(ctx context.Context, kubeClient *k8sinterface.KubernetesApi, timeout time.Duration) {
+	{
+		dynamicClient := kubeClient.DynamicClient.Resource(Schema)
+		regularClient := kubeClient.KubernetesClient.CoreV1()
+
+		for {
+			logger.L().Ctx(ctx).Info("starting a new service discovery handling")
+			currentServiceList := make(currentServiceList, 0)
```

Comment:
if capacity is already known, allocate it here to avoid redundant allocatations with append